File: delsig/calculateTF.m
function [ntf,stf] = calculateTF(ABCD,k)
% [ntf,stf] = calculateTF(ABCD,k=1) 
% Calculate the NTF and STF of a delta-sigma modulator whose loop filter
% is described by the ABCD matrix, assuming a quantizer gain of k.
% The NTF and STF are zpk objects.
if nargin < 2 | isnan(k)
    k = 1;
end
nq = length(k);
n = size(ABCD,1) - nq;
nu = size(ABCD,2) - size(ABCD,1);
[A,B,C,D] = partitionABCD(ABCD,nu+nq);
B1 = B(:,1:nu);
B2 = B(:,nu+1:end);
D1 = D(:,1:nu);
if any(any( D(:,nu+1:end) ~= 0 ))
    error('D2 must be zero');
end
K = diag(k);
% Find the noise transfer function by forming the closed-loop
% system (sys_cl) in state-space form.
Acl = A + B2*K*C;
Bcl = [B1 + B2*K*D1, B2];
Ccl = K*C;
Dcl = [K*D1 eye(nq)];
vn = sscanf(version,'%d');
if vn>6 | all(imag(ABCD)==0)	% real modulator or recent version of MATLAB
    sys_cl = ss(Acl,Bcl,Ccl,Dcl,1);
    tfs = zpk(sys_cl);
    tol = min(1e-3,max(1e-6,eps^(1/(size(ABCD,1)))));
    mtfs = minreal(tfs,tol);
    stf = mtfs(:,1:nu);
    ntf = mtfs(:,nu+1:nu+nq);
else	% quadrature modulator and old version of MATLAB
    p = eig(Acl);
    ntfz = eig(A);
    ntf = setPolesAndZeros(ntfz,p,1);
    if nargout>1
        fprintf(1,'Sorry. calculateTF cannot compute the STF for a complex modulator with this version of matlab.\n');
    end
    stf=[];
end
return

% Use a loophole to set complex poles and zeros in zpk objects 
function ztf = setPolesAndZeros(z,p,k)
    tol = 3e-5;		% tolerance for pole-zero cancellation
    Z = zpk(0,[],1,1);
    ztf = zpk([],[],k,1);
    for i = 1:length(p)
	match = abs(p(i)-z)<tol;
	if any(match)
	    f = find(match);
	    z(f(1)) = [];
	else
	    ztf = ztf/(Z-p(i));
	end
    end
    for i = 1:length(z)
	ztf = ztf*(Z-z(i));
    end
return

File: pydelsigma/_calculateTF.py
# -*- coding: utf-8 -*-
# _calculateTF.py
# Module providing the calculateTF function
# Copyright 2013 Giuseppe Venturini
# This file is part of python-deltasigma.
#
# python-deltasigma is a 1:1 Python replacement of Richard Schreier's 
# MATLAB delta sigma toolbox (aka "delsigma"), upon which it is heavily based.
# The delta sigma toolbox is (c) 2009, Richard Schreier.
#
# python-deltasigma is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# LICENSE file for the licensing terms.

"""Module providing the calculateTF() function
"""

import numpy as np
from scipy.signal import lti, ss2zpk
from ._constants import eps
from ._partitionABCD import partitionABCD
from ._utils import minreal

def calculateTF(ABCD, k=1.):
	"""Calculate the NTF and STF of a delta-sigma modulator whose loop filter
	is described by the ABCD matrix, assuming a quantizer gain of k.

	**Parameters:**

	ABCD : array_like,
               The ABCD matrix that describes the system.

	k : float, optional
            The quantizer gain. If not specified, a default value of 1 is used.

	**Returns:**

	(NTF, STF) : a tuple of two LTI objects.
	"""
	A, B, C, D = partitionABCD(ABCD)
	if B.shape[1] > 1:
		B1 = B[:, 0]
		B2 = B[:, 1]
		B1 = B1.reshape((B1.shape[0], 1)) if len(B1.shape) == 1 else B1
		B2 = B2.reshape((B2.shape[0], 1)) if len(B2.shape) == 1 else B2
	else:
		B1 = B
		B2 = B

	# Find the noise transfer function by forming the closed-loop
	# system (sys_cl) in state-space form.
	Acl = A + k * np.dot(B2, C)
	Bcl = np.hstack((B1 + k*B2*D[0, 0], B2))
	Ccl = k*C
	Dcl = np.array((k*D[0, 0], 1.))
	Dcl = Dcl.reshape((1, Dcl.shape[0])) if len(Dcl.shape) == 1 else Dcl
	tol = min(1e-3, max(1e-6, eps**(1/ABCD.shape[0])))
	# input #0 is the signal
	# input #1 is the quantization noise
	stf_p, stf_z, stf_k  = ss2zpk(Acl, Bcl, Ccl, Dcl, input=0)
	ntf_p, ntf_z, ntf_k = ss2zpk(Acl, Bcl, Ccl, Dcl, input=1)
	stf = lti(stf_p, stf_z, stf_k)
	ntf = lti(ntf_p, ntf_z, ntf_k)
	stf_min, ntf_min = minreal((stf, ntf), tol)
	return ntf_min, stf_min

def test_calculateTF():
	"""Test function for calculateTF()"""
	from ._utils import cplxpair
	ABCD = [[1.000000000000000, 0., 0., 0.044408783846879, -0.044408783846879],
	        [0.999036450096481, 0.997109907515262, -0.005777399147297, 0., 0.499759089304780],
	        [0.499759089304780, 0.999036450096481, 0.997109907515262,  0., -0.260002096136488],
	        [0,                 0,                 1.000000000000000,  0, -0.796730400347216]]
	ABCD = np.array(ABCD)
	ntf, stf = calculateTF(ABCD)
	ntf_zeros, ntf_poles = np.roots(ntf.num), np.roots(ntf.den)
	stf_zeros, stf_poles = np.roots(stf.num), np.roots(stf.den)
	mntf_poles = np.array((1.498975311463384, 1.102565142679772, 0.132677264750882))
	mntf_zeros = np.array((0.997109907515262 + 0.075972576202904j, 
	                       0.997109907515262 - 0.075972576202904j, 
	                       1.000000000000000 + 0.000000000000000j)
	                      )
	mstf_zeros = np.array((-0.999999999999996,))
	mstf_poles = np.array((1.498975311463384, 1.102565142679772, 0.132677264750882))
	# for some reason, sometimes the zeros are in different order.
	ntf_zeros, mntf_zeros = cplxpair(ntf_zeros), cplxpair(mntf_zeros)
	stf_zeros, mstf_zeros = cplxpair(stf_zeros), cplxpair(mstf_zeros)
	ntf_poles, mntf_poles = cplxpair(ntf_poles), cplxpair(mntf_poles)
	stf_poles, mstf_poles = cplxpair(stf_poles), cplxpair(mstf_poles)
	assert np.allclose(ntf_zeros, mntf_zeros, rtol=1e-5, atol=1e-8)
	assert np.allclose(ntf_poles, mntf_poles, rtol=1e-5, atol=1e-8)
	assert np.allclose(stf_zeros, mstf_zeros, rtol=1e-5, atol=1e-8)
	assert np.allclose(stf_poles, mstf_poles, rtol=1e-5, atol=1e-8)
